generators/beagle/src/main/scala/BeaglePeriphery.scala

```diff
@@ -192,42 +265,141 @@ trait HasPeripheryBeagleBundle {
+  // clocks coming from offchip
+  val single_clks = IO(Input(Vec(1, Clock())))
+  val diff_clks   = IO(Input(Vec(1, Clock())))
+
+  val bh_clk_sel     = IO(Input(Bool()))
+  val rs_clk_sel     = IO(Input(Bool()))
+  val uncore_clk_sel = IO(Input(Bool()))
+
+  val uncore_clk_out = IO(Output(Clock()))
+  val bh_clk_out     = IO(Output(Clock()))
+  val rs_clk_out     = IO(Output(Clock()))
+  val lbwif_clk_out  = IO(Output(Clock()))
+
+
+  // ------------------------------------------------
+
+  val scr_mod = outer.scr.module
+
+  // other signals
+  scr_mod.io.boot := boot
+  scr_mod.io.rst_async := rst_async
+
+  // ------------------------------------------------
+
+  // get the actual clock from the io clocks
+  require(bh_clk_sel.getWidth >= (log2Ceil(single_clks.length) + log2Ceil(diff_clks.length)), "[sys-top] must be able to select all input clocks")
+  require(rs_clk_sel.getWidth >= (log2Ceil(single_clks.length) + log2Ceil(diff_clks.length)), "[sys-top] must be able to select all input clocks")
+  require(uncore_clk_sel.getWidth >= (log2Ceil(single_clks.length) + log2Ceil(diff_clks.length)), "[sys-top] must be able to select all input clocks")
+
+  // setup uncore muxes and divider
+
+  val uncore_clk_mux = testchipip.ClockMutexMux(single_clks ++ diff_clks)
+  uncore_clk_mux.io.sel := uncore_clk_sel
+  uncore_clk_mux.io.resetAsync := rst_async
+  val uncore_clk_mux_out = uncore_clk_mux.io.clockOut
+  val uncore_clk_mux_rst = ResetCatchAndSync(uncore_clk_mux.io.clockOut, reset.asBool)
+
+  val uncore_clk_divider = withClockAndReset(uncore_clk_mux_out, uncore_clk_mux_rst) {
+    Module(new testchipip.ClockDivider(outer.scrParams.uncoreClkDivBits))
+  }
+  uncore_clk_divider.io.divisor := scr_mod.io.uncore_clk_divisor
+
+  // mux between the divided uncore and the undivided
+  val uncore_clks_pre_muxed = Seq(uncore_clk_mux_out, uncore_clk_divider.io.clockOut)
+  require(scr_mod.io.uncore_clk_pass_sel.getWidth >= log2Ceil(uncore_clks_pre_muxed.length), "[sys-top] must be able to select the uncore clocks")
+  val uncore_clk_pass_mux = testchipip.ClockMutexMux(uncore_clks_pre_muxed)
+  uncore_clk_pass_mux.io.sel := scr_mod.io.uncore_clk_pass_sel
+  uncore_clk_pass_mux.io.resetAsync := rst_async
+  val uncore_clk = uncore_clk_pass_mux.io.clockOut
+  val uncore_rst = ResetCatchAndSync(uncore_clk_pass_mux.io.clockOut, reset.asBool)
+
+  // setup tile muxes
+
+  val bh_clk_mux = testchipip.ClockMutexMux(single_clks ++ diff_clks)
+  bh_clk_mux.io.sel := bh_clk_sel
+  bh_clk_mux.io.resetAsync := rst_async
+  val bh_clk_mux_out = bh_clk_mux.io.clockOut
+  val bh_clk_mux_rst = ResetCatchAndSync(bh_clk_mux.io.clockOut, reset.asBool)
+
+  val rs_clk_mux = testchipip.ClockMutexMux(single_clks ++ diff_clks)
+  rs_clk_mux.io.sel := rs_clk_sel
+  rs_clk_mux.io.resetAsync := rst_async
+  val rs_clk_mux_out = rs_clk_mux.io.clockOut
+  val rs_clk_mux_rst = ResetCatchAndSync(rs_clk_mux.io.clockOut, reset.asBool)
+
+  // setup clock dividers
```

Comment:
There's probably a way to make this more DRY with some functions you apply to generate clocks.

Reply:
DRY? Do you mean wrap these in functions so it is easier to read? That should be easy enough to do.